From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msvcp90: Export __Wcrtomb_lk everywhere and msvcp110's _Add_vtordisp members and _GetCurrentThreadId.

__Wcrtomb_lk (_Wcrtomb without the locale lock) in msvcp60 .. msvcp120
as in msvcp140 (patch 4549); msvcp110's _Add_vtordisp1/_Add_vtordisp2
members of the stream classes, which the headers declare with empty
bodies (they only give the classes vtordisp fields), and its
details::_GetCurrentThreadId. msvcp120_app forwards.
---
 dlls/msvcp100/msvcp100.spec         |  2 +-
 dlls/msvcp110/msvcp110.spec         | 76 ++++++++++++++++++-------------------
 dlls/msvcp120/msvcp120.spec         |  2 +-
 dlls/msvcp120_app/msvcp120_app.spec |  2 +-
 dlls/msvcp60/msvcp60.spec           |  2 +-
 dlls/msvcp70/msvcp70.spec           |  2 +-
 dlls/msvcp71/msvcp71.spec           |  2 +-
 dlls/msvcp80/msvcp80.spec           |  2 +-
 dlls/msvcp90/locale.c               |  2 -
 dlls/msvcp90/misc.c                 | 22 +++++++++++
 dlls/msvcp90/msvcp90.spec           |  2 +-
 11 files changed, 68 insertions(+), 48 deletions(-)

diff --git a/dlls/msvcp100/msvcp100.spec b/dlls/msvcp100/msvcp100.spec
index 2d9a96b..23017f2 100644
--- a/dlls/msvcp100/msvcp100.spec
+++ b/dlls/msvcp100/msvcp100.spec
@@ -3914,7 +3914,7 @@
 @ cdecl _Wcscoll(ptr ptr ptr ptr ptr)
 @ cdecl _Wcsxfrm(ptr ptr ptr ptr ptr)
 # extern _Xbig
-@ stub __Wcrtomb_lk
+@ cdecl __Wcrtomb_lk(ptr long ptr ptr)
 @ cdecl towctrans(long long)
 @ cdecl wctrans(str)
 @ cdecl wctype(str)
diff --git a/dlls/msvcp110/msvcp110.spec b/dlls/msvcp110/msvcp110.spec
index 58f2979..561f3ad 100644
--- a/dlls/msvcp110/msvcp110.spec
+++ b/dlls/msvcp110/msvcp110.spec
@@ -1097,42 +1097,42 @@
 @ extern ?_7@placeholders@std@@3V?$_Ph@$06@2@A _Ph_7
 @ extern ?_8@placeholders@std@@3V?$_Ph@$07@2@A _Ph_8
 @ extern ?_9@placeholders@std@@3V?$_Ph@$08@2@A _Ph_9
-@ stub -arch=arm ?_Add_vtordisp1@?$basic_ios@DU?$char_traits@D@std@@@std@@UAAXXZ
-@ stub -arch=i386 ?_Add_vtordisp1@?$basic_ios@DU?$char_traits@D@std@@@std@@UAEXXZ
-@ stub -arch=win64 ?_Add_vtordisp1@?$basic_ios@DU?$char_traits@D@std@@@std@@UEAAXXZ
-@ stub -arch=arm ?_Add_vtordisp1@?$basic_ios@GU?$char_traits@G@std@@@std@@UAAXXZ
-@ stub -arch=i386 ?_Add_vtordisp1@?$basic_ios@GU?$char_traits@G@std@@@std@@UAEXXZ
-@ stub -arch=win64 ?_Add_vtordisp1@?$basic_ios@GU?$char_traits@G@std@@@std@@UEAAXXZ
-@ stub -arch=arm ?_Add_vtordisp1@?$basic_ios@_WU?$char_traits@_W@std@@@std@@UAAXXZ
-@ stub -arch=i386 ?_Add_vtordisp1@?$basic_ios@_WU?$char_traits@_W@std@@@std@@UAEXXZ
-@ stub -arch=win64 ?_Add_vtordisp1@?$basic_ios@_WU?$char_traits@_W@std@@@std@@UEAAXXZ
-@ stub -arch=arm ?_Add_vtordisp1@?$basic_istream@DU?$char_traits@D@std@@@std@@UAAXXZ
-@ stub -arch=i386 ?_Add_vtordisp1@?$basic_istream@DU?$char_traits@D@std@@@std@@UAEXXZ
-@ stub -arch=win64 ?_Add_vtordisp1@?$basic_istream@DU?$char_traits@D@std@@@std@@UEAAXXZ
-@ stub -arch=arm ?_Add_vtordisp1@?$basic_istream@GU?$char_traits@G@std@@@std@@UAAXXZ
-@ stub -arch=i386 ?_Add_vtordisp1@?$basic_istream@GU?$char_traits@G@std@@@std@@UAEXXZ
-@ stub -arch=win64 ?_Add_vtordisp1@?$basic_istream@GU?$char_traits@G@std@@@std@@UEAAXXZ
-@ stub -arch=arm ?_Add_vtordisp1@?$basic_istream@_WU?$char_traits@_W@std@@@std@@UAAXXZ
-@ stub -arch=i386 ?_Add_vtordisp1@?$basic_istream@_WU?$char_traits@_W@std@@@std@@UAEXXZ
-@ stub -arch=win64 ?_Add_vtordisp1@?$basic_istream@_WU?$char_traits@_W@std@@@std@@UEAAXXZ
-@ stub -arch=arm ?_Add_vtordisp2@?$basic_ios@DU?$char_traits@D@std@@@std@@UAAXXZ
-@ stub -arch=i386 ?_Add_vtordisp2@?$basic_ios@DU?$char_traits@D@std@@@std@@UAEXXZ
-@ stub -arch=win64 ?_Add_vtordisp2@?$basic_ios@DU?$char_traits@D@std@@@std@@UEAAXXZ
-@ stub -arch=arm ?_Add_vtordisp2@?$basic_ios@GU?$char_traits@G@std@@@std@@UAAXXZ
-@ stub -arch=i386 ?_Add_vtordisp2@?$basic_ios@GU?$char_traits@G@std@@@std@@UAEXXZ
-@ stub -arch=win64 ?_Add_vtordisp2@?$basic_ios@GU?$char_traits@G@std@@@std@@UEAAXXZ
-@ stub -arch=arm ?_Add_vtordisp2@?$basic_ios@_WU?$char_traits@_W@std@@@std@@UAAXXZ
-@ stub -arch=i386 ?_Add_vtordisp2@?$basic_ios@_WU?$char_traits@_W@std@@@std@@UAEXXZ
-@ stub -arch=win64 ?_Add_vtordisp2@?$basic_ios@_WU?$char_traits@_W@std@@@std@@UEAAXXZ
-@ stub -arch=arm ?_Add_vtordisp2@?$basic_ostream@DU?$char_traits@D@std@@@std@@UAAXXZ
-@ stub -arch=i386 ?_Add_vtordisp2@?$basic_ostream@DU?$char_traits@D@std@@@std@@UAEXXZ
-@ stub -arch=win64 ?_Add_vtordisp2@?$basic_ostream@DU?$char_traits@D@std@@@std@@UEAAXXZ
-@ stub -arch=arm ?_Add_vtordisp2@?$basic_ostream@GU?$char_traits@G@std@@@std@@UAAXXZ
-@ stub -arch=i386 ?_Add_vtordisp2@?$basic_ostream@GU?$char_traits@G@std@@@std@@UAEXXZ
-@ stub -arch=win64 ?_Add_vtordisp2@?$basic_ostream@GU?$char_traits@G@std@@@std@@UEAAXXZ
-@ stub -arch=arm ?_Add_vtordisp2@?$basic_ostream@_WU?$char_traits@_W@std@@@std@@UAAXXZ
-@ stub -arch=i386 ?_Add_vtordisp2@?$basic_ostream@_WU?$char_traits@_W@std@@@std@@UAEXXZ
-@ stub -arch=win64 ?_Add_vtordisp2@?$basic_ostream@_WU?$char_traits@_W@std@@@std@@UEAAXXZ
+@ cdecl -arch=arm ?_Add_vtordisp1@?$basic_ios@DU?$char_traits@D@std@@@std@@UAAXXZ(ptr) basic_ios__Add_vtordisp
+@ thiscall -arch=i386 ?_Add_vtordisp1@?$basic_ios@DU?$char_traits@D@std@@@std@@UAEXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=win64 ?_Add_vtordisp1@?$basic_ios@DU?$char_traits@D@std@@@std@@UEAAXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=arm ?_Add_vtordisp1@?$basic_ios@GU?$char_traits@G@std@@@std@@UAAXXZ(ptr) basic_ios__Add_vtordisp
+@ thiscall -arch=i386 ?_Add_vtordisp1@?$basic_ios@GU?$char_traits@G@std@@@std@@UAEXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=win64 ?_Add_vtordisp1@?$basic_ios@GU?$char_traits@G@std@@@std@@UEAAXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=arm ?_Add_vtordisp1@?$basic_ios@_WU?$char_traits@_W@std@@@std@@UAAXXZ(ptr) basic_ios__Add_vtordisp
+@ thiscall -arch=i386 ?_Add_vtordisp1@?$basic_ios@_WU?$char_traits@_W@std@@@std@@UAEXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=win64 ?_Add_vtordisp1@?$basic_ios@_WU?$char_traits@_W@std@@@std@@UEAAXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=arm ?_Add_vtordisp1@?$basic_istream@DU?$char_traits@D@std@@@std@@UAAXXZ(ptr) basic_ios__Add_vtordisp
+@ thiscall -arch=i386 ?_Add_vtordisp1@?$basic_istream@DU?$char_traits@D@std@@@std@@UAEXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=win64 ?_Add_vtordisp1@?$basic_istream@DU?$char_traits@D@std@@@std@@UEAAXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=arm ?_Add_vtordisp1@?$basic_istream@GU?$char_traits@G@std@@@std@@UAAXXZ(ptr) basic_ios__Add_vtordisp
+@ thiscall -arch=i386 ?_Add_vtordisp1@?$basic_istream@GU?$char_traits@G@std@@@std@@UAEXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=win64 ?_Add_vtordisp1@?$basic_istream@GU?$char_traits@G@std@@@std@@UEAAXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=arm ?_Add_vtordisp1@?$basic_istream@_WU?$char_traits@_W@std@@@std@@UAAXXZ(ptr) basic_ios__Add_vtordisp
+@ thiscall -arch=i386 ?_Add_vtordisp1@?$basic_istream@_WU?$char_traits@_W@std@@@std@@UAEXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=win64 ?_Add_vtordisp1@?$basic_istream@_WU?$char_traits@_W@std@@@std@@UEAAXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=arm ?_Add_vtordisp2@?$basic_ios@DU?$char_traits@D@std@@@std@@UAAXXZ(ptr) basic_ios__Add_vtordisp
+@ thiscall -arch=i386 ?_Add_vtordisp2@?$basic_ios@DU?$char_traits@D@std@@@std@@UAEXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=win64 ?_Add_vtordisp2@?$basic_ios@DU?$char_traits@D@std@@@std@@UEAAXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=arm ?_Add_vtordisp2@?$basic_ios@GU?$char_traits@G@std@@@std@@UAAXXZ(ptr) basic_ios__Add_vtordisp
+@ thiscall -arch=i386 ?_Add_vtordisp2@?$basic_ios@GU?$char_traits@G@std@@@std@@UAEXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=win64 ?_Add_vtordisp2@?$basic_ios@GU?$char_traits@G@std@@@std@@UEAAXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=arm ?_Add_vtordisp2@?$basic_ios@_WU?$char_traits@_W@std@@@std@@UAAXXZ(ptr) basic_ios__Add_vtordisp
+@ thiscall -arch=i386 ?_Add_vtordisp2@?$basic_ios@_WU?$char_traits@_W@std@@@std@@UAEXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=win64 ?_Add_vtordisp2@?$basic_ios@_WU?$char_traits@_W@std@@@std@@UEAAXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=arm ?_Add_vtordisp2@?$basic_ostream@DU?$char_traits@D@std@@@std@@UAAXXZ(ptr) basic_ios__Add_vtordisp
+@ thiscall -arch=i386 ?_Add_vtordisp2@?$basic_ostream@DU?$char_traits@D@std@@@std@@UAEXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=win64 ?_Add_vtordisp2@?$basic_ostream@DU?$char_traits@D@std@@@std@@UEAAXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=arm ?_Add_vtordisp2@?$basic_ostream@GU?$char_traits@G@std@@@std@@UAAXXZ(ptr) basic_ios__Add_vtordisp
+@ thiscall -arch=i386 ?_Add_vtordisp2@?$basic_ostream@GU?$char_traits@G@std@@@std@@UAEXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=win64 ?_Add_vtordisp2@?$basic_ostream@GU?$char_traits@G@std@@@std@@UEAAXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=arm ?_Add_vtordisp2@?$basic_ostream@_WU?$char_traits@_W@std@@@std@@UAAXXZ(ptr) basic_ios__Add_vtordisp
+@ thiscall -arch=i386 ?_Add_vtordisp2@?$basic_ostream@_WU?$char_traits@_W@std@@@std@@UAEXXZ(ptr) basic_ios__Add_vtordisp
+@ cdecl -arch=win64 ?_Add_vtordisp2@?$basic_ostream@_WU?$char_traits@_W@std@@@std@@UEAAXXZ(ptr) basic_ios__Add_vtordisp
 @ cdecl -arch=arm ?_Addcats@_Locinfo@std@@QAAAAV12@HPBD@Z(ptr long str) _Locinfo__Addcats
 @ thiscall -arch=i386 ?_Addcats@_Locinfo@std@@QAEAAV12@HPBD@Z(ptr long str) _Locinfo__Addcats
 @ cdecl -arch=win64 ?_Addcats@_Locinfo@std@@QEAAAEAV12@HPEBD@Z(ptr long str) _Locinfo__Addcats
@@ -1229,7 +1229,7 @@
 @ cdecl -arch=win64 ?_Future_error_map@std@@YAPEBDH@Z(long) _Future_error_map
 @ cdecl -arch=win32 ?_GetCombinableSize@details@Concurrency@@YAIXZ() _GetCombinableSize
 @ cdecl -arch=win64 ?_GetCombinableSize@details@Concurrency@@YA_KXZ() _GetCombinableSize
-@ stub ?_GetCurrentThreadId@details@Concurrency@@YAKXZ
+@ cdecl ?_GetCurrentThreadId@details@Concurrency@@YAKXZ() _GetCurrentThreadId
 @ cdecl ?_GetNextAsyncId@details@Concurrency@@YAIXZ() GetNextAsyncId
 @ cdecl -arch=win32 ?_Get_future_error_what@std@@YAPBDH@Z(long) _Future_error_map
 @ cdecl -arch=win64 ?_Get_future_error_what@std@@YAPEBDH@Z(long) _Future_error_map
@@ -3894,7 +3894,7 @@
 @ cdecl _Xtime_diff_to_millis(ptr)
 @ cdecl _Xtime_diff_to_millis2(ptr ptr)
 @ cdecl -ret64 _Xtime_get_ticks()
-@ stub __Wcrtomb_lk
+@ cdecl __Wcrtomb_lk(ptr long ptr ptr)
 # extern _Zero
 @ cdecl towctrans(long long)
 @ cdecl wctrans(str)
diff --git a/dlls/msvcp120/msvcp120.spec b/dlls/msvcp120/msvcp120.spec
index 27c9458..46635d6 100644
--- a/dlls/msvcp120/msvcp120.spec
+++ b/dlls/msvcp120/msvcp120.spec
@@ -3842,5 +3842,5 @@
 @ cdecl _Xtime_diff_to_millis2(ptr ptr)
 @ cdecl -ret64 _Xtime_get_ticks()
 # extern _Zero
-@ stub __Wcrtomb_lk
+@ cdecl __Wcrtomb_lk(ptr long ptr ptr)
 @ cdecl xtime_get(ptr long)
diff --git a/dlls/msvcp120_app/msvcp120_app.spec b/dlls/msvcp120_app/msvcp120_app.spec
index bd816e8..0523eac 100644
--- a/dlls/msvcp120_app/msvcp120_app.spec
+++ b/dlls/msvcp120_app/msvcp120_app.spec
@@ -3842,5 +3842,5 @@
 @ cdecl _Xtime_diff_to_millis2(ptr ptr) msvcp120._Xtime_diff_to_millis2
 @ cdecl -ret64 _Xtime_get_ticks() msvcp120._Xtime_get_ticks
 # extern _Zero
-@ stub __Wcrtomb_lk
+@ cdecl __Wcrtomb_lk(ptr long ptr ptr) msvcp120.__Wcrtomb_lk
 @ cdecl xtime_get(ptr long) msvcp120.xtime_get
diff --git a/dlls/msvcp60/msvcp60.spec b/dlls/msvcp60/msvcp60.spec
index 7e8f6bf..fbf281e 100644
--- a/dlls/msvcp60/msvcp60.spec
+++ b/dlls/msvcp60/msvcp60.spec
@@ -5686,7 +5686,7 @@
 @ cdecl _Toupper(long ptr)
 @ cdecl _Wcrtomb(ptr long ptr ptr)
 # extern _Xbig
-@ stub __Wcrtomb_lk
+@ cdecl __Wcrtomb_lk(ptr long ptr ptr)
 @ cdecl btowc(long)
 @ cdecl mbrlen(ptr long ptr)
 @ cdecl mbrtowc(ptr str long ptr)
diff --git a/dlls/msvcp70/msvcp70.spec b/dlls/msvcp70/msvcp70.spec
index 59554f1..53a211f 100644
--- a/dlls/msvcp70/msvcp70.spec
+++ b/dlls/msvcp70/msvcp70.spec
@@ -6976,7 +6976,7 @@
 @ cdecl _Wcsxfrm(ptr ptr ptr ptr ptr)
 # extern _Xbig
 # extern _Zero
-@ stub __Wcrtomb_lk
+@ cdecl __Wcrtomb_lk(ptr long ptr ptr)
 @ cdecl btowc(long)
 @ cdecl mbrlen(ptr long ptr)
 @ cdecl mbrtowc(ptr str long ptr)
diff --git a/dlls/msvcp71/msvcp71.spec b/dlls/msvcp71/msvcp71.spec
index 55829c0..7adb820 100644
--- a/dlls/msvcp71/msvcp71.spec
+++ b/dlls/msvcp71/msvcp71.spec
@@ -7099,7 +7099,7 @@
 @ cdecl _Wcsxfrm(ptr ptr ptr ptr ptr)
 # extern _Xbig
 # extern _Zero
-@ stub __Wcrtomb_lk
+@ cdecl __Wcrtomb_lk(ptr long ptr ptr)
 @ cdecl btowc(long)
 @ cdecl mbrlen(ptr long ptr)
 @ cdecl mbrtowc(ptr str long ptr)
diff --git a/dlls/msvcp80/msvcp80.spec b/dlls/msvcp80/msvcp80.spec
index 199883d..726ff93 100644
--- a/dlls/msvcp80/msvcp80.spec
+++ b/dlls/msvcp80/msvcp80.spec
@@ -7913,7 +7913,7 @@
 @ stub __Stodx
 @ stub __Stofx
 @ stub __Stoldx
-@ stub __Wcrtomb_lk
+@ cdecl __Wcrtomb_lk(ptr long ptr ptr)
 @ cdecl towctrans(long long)
 @ cdecl wctrans(str)
 @ cdecl wctype(str)
diff --git a/dlls/msvcp90/locale.c b/dlls/msvcp90/locale.c
index 2682614..d618e5c 100644
--- a/dlls/msvcp90/locale.c
+++ b/dlls/msvcp90/locale.c
@@ -5067,7 +5067,6 @@ int __cdecl _Wcrtomb(char *s, wchar_t wch, _Mbstatet *state, const _Cvtvec *cvt)
     return size;
 }
 
-#if _MSVCP_VER >= 140
 /* __Wcrtomb_lk: _Wcrtomb, without taking the locale lock */
 int __cdecl __Wcrtomb_lk(char *s, wchar_t wch, _Mbstatet *state, const _Cvtvec *cvt)
 {
@@ -5076,7 +5075,6 @@ int __cdecl __Wcrtomb_lk(char *s, wchar_t wch, _Mbstatet *state, const _Cvtvec *
 #endif
     return _Wcrtomb(s, wch, state, cvt);
 }
-#endif
 
 /* ?_Donarrow@?$ctype@_W@std@@IBED_WD@Z */
 /* ?_Donarrow@?$ctype@_W@std@@IEBAD_WD@Z */
diff --git a/dlls/msvcp90/misc.c b/dlls/msvcp90/misc.c
index 3b7f88c..2386edd 100644
--- a/dlls/msvcp90/misc.c
+++ b/dlls/msvcp90/misc.c
@@ -2263,3 +2263,25 @@ void __cdecl _Once(LONG *flag, void (__cdecl *func)(void))
     while (ReadAcquire(flag) != 2) Sleep(0);
 }
 #endif
+
+#if _MSVCP_VER == 110
+/* ?_Add_vtordisp1@?$basic_ios@DU?$char_traits@D@std@@@std@@UAEXXZ */
+/* ?_Add_vtordisp2@?$basic_ostream@DU?$char_traits@D@std@@@std@@UAEXXZ */
+/* and their wchar_t and unsigned short forms: virtual functions without a
+ * body, there only to give the stream classes vtordisp fields */
+DEFINE_THISCALL_WRAPPER(basic_ios__Add_vtordisp, 4)
+void __thiscall basic_ios__Add_vtordisp(void *this)
+{
+    TRACE("(%p)\n", this);
+}
+
+/* ?_GetCurrentThreadId@details@Concurrency@@YAKXZ */
+DWORD __cdecl _GetCurrentThreadId(void)
+{
+    TRACE("()\n");
+#ifdef SG_MUTANT_CONCRT_THREAD_ID_PROCESS
+    return GetCurrentProcessId();
+#endif
+    return GetCurrentThreadId();
+}
+#endif
diff --git a/dlls/msvcp90/msvcp90.spec b/dlls/msvcp90/msvcp90.spec
index 9750a75..cccfac4 100644
--- a/dlls/msvcp90/msvcp90.spec
+++ b/dlls/msvcp90/msvcp90.spec
@@ -7937,7 +7937,7 @@
 @ cdecl _Wcscoll(ptr ptr ptr ptr ptr)
 @ cdecl _Wcsxfrm(ptr ptr ptr ptr ptr)
 # extern _Xbig
-@ stub __Wcrtomb_lk
+@ cdecl __Wcrtomb_lk(ptr long ptr ptr)
 @ cdecl towctrans(long long)
 @ cdecl wctrans(str)
 @ cdecl wctype(str)
